File: client/src/boundary/interactor/GetGameInteractor.test.ts
import GameStorage from "services/api/GameStorage";
import { GameD2BConverter } from "./GameD2BConverter";
import { GetGameInteractor } from "./GetGameInteractor";
import { Board, Game } from "domain/index";
import { BoundaryBoard, BoundaryGame } from "boundary/model/index";
import { mock, MockProxy } from "jest-mock-extended";
import { of } from "rxjs";
import { GameState } from "utils/Constants";

describe("Get Game Interactor", () => {
  let gameStorage: MockProxy<GameStorage>;
  let gameD2BConverter: MockProxy<GameD2BConverter>;
  let getGameInteractor: GetGameInteractor;
  const inputId = "123";

  beforeEach(() => {
    gameStorage = mock<GameStorage>();
    gameD2BConverter = mock<GameD2BConverter>();
    getGameInteractor = new GetGameInteractor(gameStorage, gameD2BConverter);
  });

  it("finds game and returns it", (done) => {
    const expectedGame: Game = new Game(
      "123",
      GameState.IN_PROGRESS,
      mock<Board>(),
      25,
      0
    );
    const expectedBoundaryGame: BoundaryGame = new BoundaryGame(
      "123",
      GameState.IN_PROGRESS,
      mock<BoundaryBoard>()
    );

    gameStorage.getGame.calledWith(inputId).mockReturnValue(of(expectedGame));
    gameD2BConverter.convert
      .calledWith(expectedGame)
      .mockReturnValue(expectedBoundaryGame);

    getGameInteractor.getGame(inputId).subscribe({
      next: (game) => {
        expect(game).toStrictEqual(expectedBoundaryGame);
        done();
      },
      error: (error) => done(error),
    });
  });
});

File: server/battleships/use_cases/implementation/src/test/java/eu/optas/use_cases/implementation/GetGameInteractorTest.java
package eu.optas.use_cases.implementation;

import eu.optas.domain.Board;
import eu.optas.domain.Game;
import eu.optas.gateway.api.GameGateway;
import eu.optas.use_cases.api.BoundaryBoard;
import eu.optas.use_cases.api.BoundaryGame;
import eu.optas.utils.GameState;
import org.junit.jupiter.api.Test;

import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

class GetGameInteractorTest {

    public static final String GAME_ID = "123";
    public static final GameGateway GAME_GATEWAY_MOCK = mock(GameGateway.class);
    public static final GameD2BConverter GAME_D_2_B_CONVERTER_MOCK = mock(GameD2BConverter.class);
    public static final GetGameInteractor GET_GAME_INTERACTOR =
            new GetGameInteractor(GAME_GATEWAY_MOCK, GAME_D_2_B_CONVERTER_MOCK);

    @Test
    void getGame() {
        Game expectedGame = new Game(
                GAME_ID,
                GameState.IN_PROGRESS,
                mock(Board.class),
                25,
                0
        );
        BoundaryGame expectedBoundaryGame = new BoundaryGame(
                GAME_ID,
                GameState.IN_PROGRESS,
                mock(BoundaryBoard.class),
                25,
                0
        );

        when(GAME_GATEWAY_MOCK.getGame(GAME_ID)).thenReturn(expectedGame);
        when(GAME_D_2_B_CONVERTER_MOCK.convert(expectedGame)).thenReturn(expectedBoundaryGame);

        BoundaryGame returnedGame = GET_GAME_INTERACTOR.getGame(GAME_ID).get();

        assertThat(returnedGame).usingRecursiveComparison().isEqualTo(expectedBoundaryGame);
    }

    @Test
    void getGameNull() {
        when(GAME_GATEWAY_MOCK.getGame(GAME_ID)).thenReturn(null);

        Optional<BoundaryGame> returnedGame = GET_GAME_INTERACTOR.getGame(GAME_ID);

        assertThat(returnedGame.isEmpty()).isTrue();
    }
}